Geometry/HGCalMapping/plugins/HGCalMappingESProducer.cc

```diff
@@ -102,15 +103,26 @@ void HGCalMappingESProducer::prepareModuleMapperIndexer() {
   auto& pmap = parsedMaps_["modules"];
   auto& entities = pmap.getEntries();
   for (auto row : entities) {
-    std::string typecode = pmap.getAttr("typecode", row);
-
-    if (typecode.find('M') == 0 && typecode.size() > 4)
-      typecode = typecode.substr(0, 4);
+    std::string typecode = pmap.getAttr("typecode", row);  // module type code
+    std::string wtypecode;                                 // wafer type code
+
+    // match module type code to regular expression pattern (MM-TTTT-LL-NNNN)
+    // see https://edms.cern.ch/ui/#!master/navigator/document?D:101059405:101148061:subDocs
+    //const std::regex typecode_regex("([MX])([LH])-([FTBLR5])([123])([WPC])-([A-Z]{2})-([0-9]{3,4})"); // MM-TTTT-LL-NNNN
+    const std::regex typecode_regex("(([MX])([LH])-([FTBLR5])).*");  // MM-T*
+    std::smatch typecode_match;                                      // match object for string objects
+    bool matched = std::regex_match(typecode, typecode_match, typecode_regex);
+    if (matched) {
+      wtypecode = typecode_match[1].str();  // wafer type following MM-T pattern, e.g. "MH-F"
+    } else {
+      edm::LogWarning("HGCalMappingIndexESSource")
+          << "Could not match module type code to expected pattern: " << typecode;
+    }
 
     try {
-      typecodeidx = cellIndexer_.getEnumFromTypecode(typecode);
-      nwords = cellIndexer_.getNWordsExpectedFor(typecode);
-      nerx = cellIndexer_.getNErxExpectedFor(typecode);
+      typecodeidx = cellIndexer_.getEnumFromTypecode(wtypecode);
+      nwords = cellIndexer_.getNWordsExpectedFor(wtypecode);
+      nerx = cellIndexer_.getNErxExpectedFor(wtypecode);
     } catch (cms::Exception& e) {
```

Comment:
For a subsequent PR (since the pattern exists already): it would be much better to do these searches without using exceptions for communicating the element was not found (apparently them being missing is not that exceptional). The functions could return e.g. `std::optional`. (if there is a use case where the exception should propagate to the framework, there could be another set of functions for the use case here)

Reply:
I leave it is as it is. We don't yet have all the different module readout maps hence the reason of this fallback. I don't think there is need to propagate it higher in the FWK as it's a minor exception which has a simple fix: allocate the max. memory (=max.channels in a HGCAL module) possible in case an unknown module type is parsed.

Reply:
To me the situation (failed searches) sounds like using exceptions for control flow, which is a discouraged pattern. It will make debugging of other exceptions (in a job that uses `HGCalMappingESProducer`) more difficult.

As the structure already exists, I think we can tolerate to keep it in this PR, but the exceptions really should be removed in a subsequent PR.